### PR TITLE
fix: asTokenId can generate TokenId with negative shard and realm

### LINKED ISSUE #18926: asTokenId can generate TokenId with negative shard and realm
### Background

In the following code:

```
public static TokenID asTokenId(@NonNull final com.esaulpaugh.headlong.abi.Address address) {
    final var explicit = explicitFromHeadlong(address);
    return TokenID.newBuilder()
            .shardNum(shardOfLongZero(explicit))
            .realmNum(realmOfLongZero(explicit))
            .tokenNum(numberOfLongZero(explicit))
            .build();
}
```
shardOfLongZero, realmOfLongZero and numberOfLongZero can return negative numbers if the input address/hex value doesn’t fit in a signed long(-2^63 to 2^63 - 1) ,behaviour that was observed during tests in the mirror-node

### Acceptance Criteria

1.Add check that input address fit in a long (-2^63 to 2^63 - 1) 

### Dependencies

_No response_

### Definition of Ready (DoR) Checklist

- [ ] Clear acceptance criteria
- [ ] Clear and detailed description
- [ ] Dependencies identified
- [ ] Links to documentation
- [ ] Should be completable in 2-3 Days
- [ ] Initial draft of Low-level design document
- [ ] At least high level test plan
- [ ] Groomed/Estimated

### Definition of Done (DoD) Checklist

- [ ] Acceptance Criteria complete
- [ ] No Codacy issues greater than minor (in new code)
- [ ] JavaDocs updated/created
- [ ] Code commented
- [ ] Unit tests created/updated
- [ ] 80% test code coverage (in new code)
- [ ] Happy Path and major negative cases in HAPI tests as applicable

### PR DESCRIPTION
**Pull request by vtronkov**
**Description**:
* Add validation for `asTokenId` to check if the shard, realm or number are negative values
* Add unit tests
------
* Fixing a bug in the 4-byte `longFrom` method that was not consistent with the 8-byte method.
How does the 8-byte `longFrom` method work?
- `& 0xFFL` with each param, which essentially treats the value as an unsigned value
- Shifts the value to the leftmost group of bits
- The first bytes are shifted 56 positions, making the first byte reside in the most left binary long position. Meaning that from this group of bytes, if the leftmost bit is 0, this is a positive value, if it's 1, it is a negative value.
That's why the method can return a negative value

The 4-byte `longFrom` method has a subtle bug:
- because it's returning a long(8 bytes) and the first byte is always shifted 24 positions, we are never going to change the left-most bit, meaning that it will always return a positive value.

The fix is to return int(4 bytes) instead of long(8 bytes). That will make it possible to returna  negative value because we have access to the leftmost bit.

**Related issue(s)**:

Fixes #18926

### CODE REVIEW

File: hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/utils/ConversionUtilsTest.java [hunk 1]
```diff
@@ -314,6 +314,47 @@ void evmContractIDToNumNonLongZeroTest() {
...
+    }
+
+    @Test
+    void asTokenIdWithNegativeTokenNumer() {
+        final var address = com.esaulpaugh.headlong.abi.Address.wrap("0x000000050000000000000006ffFFFFfFFffFFFff");
+        assertThrows(IllegalArgumentException.class, () -> ConversionUtils.asTokenId(address));
+    }
+
+    @Test
+    void asTokenIdWithNegativeRealm() {
```
Comment (mhess-swl) [inline hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/utils/ConversionUtilsTest.java:346]: Would be good to add a test for negative shard as well. Could be done in a future PR

**Review by mhess-swl — APPROVED** [review]

File: hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/utils/ConversionUtilsTest.java (hunk 1 above)
Comment (mhess-swl) [reply hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/utils/ConversionUtilsTest.java:346]: Weird...would be good to dig in and figure this out sometime

**Comment by vtronkov** [comment]
@mhess-swl notice my latest changes to `AssociatePrecompileSuite` and `DissociatePrecompileSuite`. Before we were doing the dispatch and on `handle` we are failing and a record was created. Now we are failing before the dispatch and no record is created. Do you think this is ok?

cc: @Neeharika-Sompalli @tinker-michaelj

**Comment by mhess-swl** [comment]
> @mhess-swl notice my latest changes to `AssociatePrecompileSuite` and `DissociatePrecompileSuite`. Before we were doing the dispatch and on `handle` we are failing and a record was created. Now we are failing before the dispatch and no record is created. Do you think this is ok?
> 
> cc: @Neeharika-Sompalli @tinker-michaelj

I would classify this behavior as a bug in the code. I suppose it _could_ be considered a breaking API change–it's theoretically possible that some application in the wild could indirectly depend on this behavior–but realistically this seems pretty unlikely.

File: hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/utils/ConversionUtilsTest.java (hunk 1 above)
Comment (vtronkov) [reply hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/utils/ConversionUtilsTest.java]: @mhess-swl I dug deeper and I found a subtle bug. I fixed it with my last commit and added a detailed description in this PR's description.

**Review by mhess-swl — APPROVED** [review]

**Review by stoyanov-st — APPROVED** [review]
LGTM,
Thanks @vtronkov !
